various-receipes/typescript/test/typescript-example.ts

```diff
@@ -0,0 +1,257 @@
+
+      // ️️️✅ Best Practice: Intercept requests for 3rd party services to eliminate undesired side effects like emails or SMS
+      // ️️️✅ Best Practice: Specify the body when you need to make sure you call the 3rd party service as expected
+      let emailPayload;
+      nock('http://mailer.com')
+        .post('/send', (payload) => ((emailPayload = payload), true))
+        .reply(202);
+
+      const orderToAdd = {
+        userId: 1,
+        productId: 2,
+        mode: 'approved',
+      };
+
+      //Act
+      await request(expressApp).post('/order').send(orderToAdd);
+
+      //Assert
+      // ️️️✅ Best Practice: Assert that the app called the mailer service appropriately
+      expect(emailPayload).toMatchObject({
+        subject: expect.any(String),
+        body: expect.any(String),
+        recipientAddress: expect.stringMatching(
+          /^[\w-\.]+@([\w-]+\.)+[\w-]{2,4}$/
+        ),
+      });
+    });
+
+    // ️️️✅ Best Practice: Check invalid input
+    test('When adding an order without specifying product, stop and return 400', async () => {
+      //Arrange
+      const orderToAdd = {
+        userId: 1,
+        mode: 'draft',
+      };
+
+      //Act
+      const orderAddResult = await request(expressApp)
+        .post('/order')
+        .send(orderToAdd);
+
+      //Assert
+      expect(orderAddResult.status).toBe(400);
+    });
+
+    // ️️️✅ Best Practice: Check error handling
+    test.todo('When a new order failed, an invalid-order error was handled');
+
+    // ️️️✅ Best Practice: Check monitoring metrics
+    test.todo(
+      'When a new valid order was added, then order-added metric was fired'
+    );
+
+    // ️️️✅ Best Practice: Simulate external failures
+    test.todo(
+      'When the user service is down, then order is still added successfully'
+    );
+
+    test('When the user does not exist, return 404 response', async () => {
+      //Arrange
+      nock('http://localhost/user/').get(`/7`).reply(404, {
+        message: 'User does not exist',
+        code: 'nonExisting',
+      });
+      const orderToAdd = {
+        userId: 7,
+        productId: 2,
+        mode: 'draft',
+      };
+
+      //Act
+      const orderAddResult = await request(expressApp)
+        .post('/order')
+        .send(orderToAdd);
+
+      //Assert
+      expect(orderAddResult.status).toBe(404);
+    });
+
+    test('When order failed, send mail to admin', async () => {
+      //Arrange
+      process.env.SEND_MAILS = 'true';
+      // ️️️✅ Best Practice: Intercept requests for 3rd party services to eliminate undesired side effects like emails or SMS
+      // ️️️✅ Best Practice: Specify the body when you need to make sure you call the 3rd party service as expected
+      let emailPayload;
+      nock('http://mailer.com')
+        .post('/send', (payload) => ((emailPayload = payload), true))
+        .reply(202);
+
+      sinon
+        .stub(OrderRepository.prototype, 'addOrder')
```

Comment:
Can we stub in a type-safe way? It means that the compiler won't allow us to override non-existing methods and params

If not achievable with sinon, there are other TD libs that are dedicated to TS